File: backend/utils/encryption.py
import base64
import os
import struct

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
import logging

logger = logging.getLogger(__name__)

# Load the master secret from environment variables. This must be a securely managed 32-byte key.
ENCRYPTION_SECRET = os.getenv('ENCRYPTION_SECRET', '').encode('utf-8')
if not ENCRYPTION_SECRET or len(ENCRYPTION_SECRET) < 32:
    raise ValueError(
        "ENCRYPTION_SECRET environment variable not set or is too short. " "It must be a securely managed 32-byte key."
    )


def derive_key(uid: str) -> bytes:
    """
    Derives a user-specific 32-byte key from the master secret and user ID (salt).
    """
    hkdf = HKDF(
        algorithm=hashes.SHA256(),
        length=32,
        salt=uid.encode('utf-8'),
        info=b'user-data-encryption',
    )
    return hkdf.derive(ENCRYPTION_SECRET)


def encrypt(data: str, uid: str) -> str:
    """
    Encrypts a string using a user-specific key.
    Returns a base64 encoded string containing nonce + ciphertext + tag.
    """
    if not data:
        return data
    key = derive_key(uid)
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)  # GCM standard nonce size

    # Data must be bytes
    plaintext_bytes = data.encode('utf-8')

    ciphertext = aesgcm.encrypt(nonce, plaintext_bytes, None)

    # Combine nonce and ciphertext for storage
    encrypted_payload = nonce + ciphertext

    return base64.b64encode(encrypted_payload).decode('utf-8')


def decrypt(encrypted_data: str, uid: str) -> str:
    """
    Decrypts a base64 encoded string using a user-specific key.
    """
    if not encrypted_data or not isinstance(encrypted_data, str):
        return encrypted_data

    try:
        key = derive_key(uid)
        aesgcm = AESGCM(key)

        encrypted_payload = base64.b64decode(encrypted_data.encode('utf-8'))

        # Extract nonce and ciphertext
        nonce = encrypted_payload[:12]
        ciphertext = encrypted_payload[12:]

        decrypted_bytes = aesgcm.decrypt(nonce, ciphertext, None)

        return decrypted_bytes.decode('utf-8')
    except Exception as e:
        # If decryption fails (e.g., wrong key, corrupted data), return the original encrypted data
        # to avoid data loss and to make debugging easier. In a production system, you might want
        # to log this error.
        logger.info(f"Decryption failed for user {uid}: {e}")
        return encrypted_data


def encrypt_audio_chunk(data: bytes, uid: str) -> bytes:
    """
    Encrypt audio chunk and return length-prefixed binary format.
    Format: [4 bytes length][12 bytes nonce][ciphertext + tag]

    This format allows concatenating multiple encrypted chunks without decryption.
    """
    key = derive_key(uid)
    aesgcm = AESGCM(key)
    nonce = os.urandom(12)

    # Encrypt (includes authentication tag)
    ciphertext = aesgcm.encrypt(nonce, data, None)

    # Combine nonce + ciphertext
    encrypted_payload = nonce + ciphertext

    # Add length prefix (4 bytes, big-endian)
    length = len(encrypted_payload)
    return struct.pack('>I', length) + encrypted_payload


def decrypt_audio_chunk(encrypted_data: bytes, uid: str, offset: int = 0):
    """
    Decrypt a single length-prefixed chunk.
    Returns: (decrypted_data, bytes_consumed)
    """
    # Read length prefix
    length = struct.unpack('>I', encrypted_data[offset : offset + 4])[0]
    offset += 4

    # Extract encrypted payload
    encrypted_payload = encrypted_data[offset : offset + length]

    # Extract nonce and ciphertext
    nonce = encrypted_payload[:12]
    ciphertext = encrypted_payload[12:]

    # Decrypt
    key = derive_key(uid)
    aesgcm = AESGCM(key)
    decrypted = aesgcm.decrypt(nonce, ciphertext, None)

    return decrypted, 4 + length


def decrypt_audio_file(encrypted_data: bytes, uid: str) -> bytes:
    """
    Decrypt an entire merged audio file (multiple concatenated chunks).
    Each chunk is length-prefixed, allowing simple concatenation during merge.
    """
    decrypted_audio = bytearray()
    offset = 0

    while offset < len(encrypted_data):
        chunk_data, bytes_consumed = decrypt_audio_chunk(encrypted_data, uid, offset)
        decrypted_audio.extend(chunk_data)
        offset += bytes_consumed

    return bytes(decrypted_audio)

File: edge/packages/platform/src/encryption.ts
/**
 * Per-user AES-256-GCM + HKDF-SHA256 — wire-compatible with backend/utils/encryption.py
 *
 * String payload: base64(nonce12 || ciphertext||tag)
 * Audio chunk: be32(len) || nonce12 || ciphertext||tag
 */

const INFO = new TextEncoder().encode("user-data-encryption");

function requireSecret(secret: string): Uint8Array {
  const bytes = new TextEncoder().encode(secret);
  if (bytes.length < 32) {
    throw new Error("ENCRYPTION_SECRET not set or too short (need >= 32 bytes)");
  }
  return bytes;
}

export async function deriveKey(secret: string, uid: string): Promise<CryptoKey> {
  const master = requireSecret(secret);
  const baseKey = await crypto.subtle.importKey("raw", master, "HKDF", false, ["deriveKey"]);
  return crypto.subtle.deriveKey(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: new TextEncoder().encode(uid),
      info: INFO,
    },
    baseKey,
    { name: "AES-GCM", length: 256 },
    true,
    ["encrypt", "decrypt"],
  );
}

export async function deriveKeyRaw(secret: string, uid: string): Promise<Uint8Array> {
  const key = await deriveKey(secret, uid);
  return new Uint8Array(await crypto.subtle.exportKey("raw", key));
}

function b64Encode(bytes: Uint8Array): string {
  let s = "";
  for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]!);
  return btoa(s);
}

function b64Decode(data: string): Uint8Array {
  const bin = atob(data);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export async function encrypt(data: string, uid: string, secret: string): Promise<string> {
  if (!data) return data;
  const key = await deriveKey(secret, uid);
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const plaintext = new TextEncoder().encode(data);
  const ciphertext = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce }, key, plaintext),
  );
  const payload = new Uint8Array(nonce.length + ciphertext.length);
  payload.set(nonce, 0);
  payload.set(ciphertext, nonce.length);
  return b64Encode(payload);
}

export async function decrypt(encryptedData: string, uid: string, secret: string): Promise<string> {
  if (!encryptedData) return encryptedData;
  try {
    const key = await deriveKey(secret, uid);
    const payload = b64Decode(encryptedData);
    const nonce = payload.slice(0, 12);
    const ciphertext = payload.slice(12);
    const plain = await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce }, key, ciphertext);
    return new TextDecoder().decode(plain);
  } catch {
    // Match Python: fail-open return ciphertext on decrypt error
    return encryptedData;
  }
}

export async function encryptAudioChunk(data: Uint8Array, uid: string, secret: string): Promise<Uint8Array> {
  const key = await deriveKey(secret, uid);
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const ciphertext = new Uint8Array(
    await crypto.subtle.encrypt({ name: "AES-GCM", iv: nonce }, key, data),
  );
  const encryptedPayload = new Uint8Array(nonce.length + ciphertext.length);
  encryptedPayload.set(nonce, 0);
  encryptedPayload.set(ciphertext, nonce.length);
  const out = new Uint8Array(4 + encryptedPayload.length);
  const view = new DataView(out.buffer);
  view.setUint32(0, encryptedPayload.length, false);
  out.set(encryptedPayload, 4);
  return out;
}

export async function decryptAudioChunk(
  encryptedData: Uint8Array,
  uid: string,
  secret: string,
  offset = 0,
): Promise<{ data: Uint8Array; bytesConsumed: number }> {
  const view = new DataView(encryptedData.buffer, encryptedData.byteOffset, encryptedData.byteLength);
  const length = view.getUint32(offset, false);
  const start = offset + 4;
  const encryptedPayload = encryptedData.slice(start, start + length);
  const nonce = encryptedPayload.slice(0, 12);
  const ciphertext = encryptedPayload.slice(12);
  const key = await deriveKey(secret, uid);
  const plain = new Uint8Array(
    await crypto.subtle.decrypt({ name: "AES-GCM", iv: nonce }, key, ciphertext),
  );
  return { data: plain, bytesConsumed: 4 + length };
}

export async function decryptAudioFile(
  encryptedData: Uint8Array,
  uid: string,
  secret: string,
): Promise<Uint8Array> {
  const chunks: Uint8Array[] = [];
  let offset = 0;
  let total = 0;
  while (offset < encryptedData.length) {
    const { data, bytesConsumed } = await decryptAudioChunk(encryptedData, uid, secret, offset);
    chunks.push(data);
    total += data.length;
    offset += bytesConsumed;
  }
  const out = new Uint8Array(total);
  let o = 0;
  for (const c of chunks) {
    out.set(c, o);
    o += c.length;
  }
  return out;
}
